simd/src/Kokkos_Simd_Scalar.hpp

```diff
@@ -0,0 +1,353 @@
+}
+
+template <class T>
+class const_where_expression<simd_mask<T, simd_abi::scalar>,
+                             simd<T, simd_abi::scalar>> {
+ public:
+  using abi_type   = simd_abi::scalar;
+  using value_type = simd<T, abi_type>;
+  using mask_type  = simd_mask<T, abi_type>;
+
+ protected:
+  value_type& m_value;
+  mask_type const& m_mask;
+
+ public:
+  KOKKOS_FORCEINLINE_FUNCTION
+  const_where_expression(mask_type const& mask_arg, value_type const& value_arg)
+      : m_value(const_cast<value_type&>(value_arg)), m_mask(mask_arg) {}
+  KOKKOS_FORCEINLINE_FUNCTION
+  mask_type const& mask() const { return m_mask; }
+  KOKKOS_FORCEINLINE_FUNCTION
+  value_type const& value() const { return m_value; }
+  KOKKOS_FORCEINLINE_FUNCTION
+  void copy_to(T* mem, element_aligned_tag) const {
+    if (static_cast<bool>(m_mask)) *mem = static_cast<T>(m_value);
+  }
+  template <class Integral>
+  KOKKOS_FORCEINLINE_FUNCTION std::enable_if_t<std::is_integral_v<Integral>>
+  scatter_to(T* mem, simd<Integral, simd_abi::scalar> const& index) const {
+    if (static_cast<bool>(m_mask))
+      mem[static_cast<Integral>(index)] = static_cast<T>(m_value);
+  }
+};
+
+template <class T>
+class where_expression<simd_mask<T, simd_abi::scalar>,
+                       simd<T, simd_abi::scalar>>
+    : public const_where_expression<simd_mask<T, simd_abi::scalar>,
+                                    simd<T, simd_abi::scalar>> {
+  using base_type = const_where_expression<simd_mask<T, simd_abi::scalar>,
+                                           simd<T, simd_abi::scalar>>;
+
+ public:
+  using typename base_type::value_type;
+  KOKKOS_FORCEINLINE_FUNCTION
+  where_expression(simd_mask<T, simd_abi::scalar> const& mask_arg,
+                   simd<T, simd_abi::scalar>& value_arg)
+      : base_type(mask_arg, value_arg) {}
+  KOKKOS_FORCEINLINE_FUNCTION
+  void copy_from(T const* mem, element_aligned_tag) {
+    if (static_cast<bool>(this->m_mask)) this->m_value = *mem;
+  }
+  template <class Integral>
+  KOKKOS_FORCEINLINE_FUNCTION std::enable_if_t<std::is_integral_v<Integral>>
+  gather_from(T const* mem, simd<Integral, simd_abi::scalar> const& index) {
+    if (static_cast<bool>(this->m_mask))
+      this->m_value = mem[static_cast<Integral>(index)];
+  }
+  template <class U>
+  KOKKOS_FORCEINLINE_FUNCTION void operator=(U const& other) {
+    if (static_cast<bool>(this->m_mask))
+      this->m_value = static_cast<simd<T, simd_abi::scalar>>(other);
+  }
+};
+
+template <class T, class BinaryOp>
+[[nodiscard]] KOKKOS_FORCEINLINE_FUNCTION T
+reduce(const_where_expression<simd_mask<T, simd_abi::scalar>,
+                              simd<T, simd_abi::scalar>> const& x,
+       T identity_element, BinaryOp) {
+  return static_cast<bool>(x.mask()) ? static_cast<T>(x.value())
+                                     : identity_element;
+}
+
+// these are specialized because NumericTraits is in KokkosKernels not Kokkos
```

Comment:
What do you mean?

Reply:
Ideally these would use `std::numeric_traits<T>::max()` and friends. On CUDA this isn't possible unless you enable experimental `constexpr` support in CUDA. The comment refers to my desire to have `Kokkos::numeric_traits` in `kokkoscore` and use it here.